.circleci/config.yml

```diff
@@ -1,173 +1,193 @@
-            only: master
+      # publish jobs depend on this as well,
+      # thus tags need to be allowed for these
+      - lint: {filters: {<<: *tags}}
+      - test: {filters: {<<: *tags}}
+
+      - build/loki:
+          requires: [ lint, test ]
+          filters: {<<: *no-master}
+      - publish/loki:
+          requires: [ lint, test ]
+          filters: { <<: *tag-or-master }
+
+      - build/canary:
+          requires: [ lint, test ]
+          filters: {<<: *no-master}
+      - publish/canary:
+          requires: [ lint, test ]
+          filters: { <<: *tag-or-master }
+
+      - build/promtail:
+          requires: [ lint, test ]
+          filters: {<<: *no-master}
+      - publish/promtail:
+          requires: [ lint, test ]
+          filters: { <<: *tag-or-master }
+
+      - build/docker-driver:
+          requires: [ lint, test ]
+          filters: {<<: *no-master}
+      - publish/docker-driver:
+          requires: [ lint, test ]
+          filters: { <<: *tag-or-master }
+    
+      - test-helm:
+          requires: [ lint, test ]
+          filters: {<<: *tags}
+      - publish-helm:
+          requires: [ test-helm ]
+          filters: {<<: *tag-or-master}
+
 
 # https://circleci.com/blog/circleci-hacks-reuse-yaml-in-your-circleci-config-with-yaml/
-defaults: &defaults
+.defaults: &defaults
   docker:
     - image: grafana/loki-build-image:0.3.0
   working_directory: /go/src/github.com/grafana/loki
 
+.machine: &machine
+  machine:
+    image: ubuntu-1604:201903-01
+  working_directory: ~/go/src/github.com/grafana/loki
+  environment:
+    APP: to-be-set
+    GOPATH: /home/circleci/go
+
+.rootless: &rootless
+  run:
+    name: rootless
+    command: |
+      sudo apt-get install -qy uidmap libseccomp-dev binfmt-support go-bindata
+      sudo docker run --privileged linuxkit/binfmt:v0.6
+
+.img: &img
+  run:
+    name: img
+    # TODO: switch to https://github.com/genuinetools/img once 5a8119fb4ce7d712ca2ed589a345213fdf576268 is released
+    command: |
+      sudo curl -fSL "https://github.com/sh0rez/img/releases/download/v0.5.8/img-linux-amd64" -o "/usr/local/bin/img"
+      sudo chmod a+x "/usr/local/bin/img"
+
+# builds a container
+.container: &container
+  <<: *machine
+  steps:
+    - checkout
+    - <<: *rootless
+    - <<: *img
+    - run:
+        name: container
+        command: |
+          make $APP-image
+
+# builds and pushes a container
+.publish: &publish
+  <<: *machine
+  steps:
+    - checkout
+    - <<: *rootless
+    - <<: *img
+    - run:
+        name: login
+        command: img login -u "$DOCKER_USER" -p "$DOCKER_PASS"
+    - run:
+        name: container
+        command: |
+          make $APP-image
+    - run:
+        name: push image
+        command: make $APP-push
+
 jobs:
   test:
     <<: *defaults
     steps:
       - checkout
-
       - run:
-          name: Run Unit Tests
-          command: |
-            make test
+          name: Unit Tests
+          command: make test
```

Comment:
I need to fix this in the current config.yaml but this should be `make BUILD_IN_CONTAINER=false test` because `test` has a pre-req on `all` which will build files and we are already running from within the build image